src/stdlib_str2num.fypp

```diff
@@ -0,0 +1,553 @@
+    end function
+    
+    function to_quad_p(s,mold,stat) result(r)
+        ! -- In/out Variables
+        character(len=:), pointer :: s !> input string
+        real(qp), intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        real(qp) :: r    !> Output real value
+        integer(1),intent(inout), optional :: stat
+        ! -- Internal Variables
+        integer(1) :: p !> position within the number
+        integer(1) :: err
+        !----------------------------------------------
+        call to_num_base(s,r,p,err)
+        p = min( p , len(s) )
+        s => s(p:)
+        if(present(stat)) stat = err
+    end function
+#:endif
+
+    !---------------------------------------------
+    ! String To Number Implementations
+    !---------------------------------------------
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    elemental subroutine to_int_${k1}$(s,v,p,stat)
+        !> Return an unsigned 32-bit integer
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        ${t1}$, intent(out)  :: v !> Output real value
+        integer(int8), intent(out)  :: p !> position within the number
+        integer(int8), intent(out)  :: stat !> status upon succes or failure to read
+        ! -- Internal Variables
+        integer(int8)  :: val 
+        !----------------------------------------------
+        stat = 23 !> initialize error status with any number > 0
+        !----------------------------------------------
+        ! Find first non white space
+        p = mvs2nwsp(s)
+        !----------------------------------------------
+        v = 0
+        do while( p<=len(s) )
+            val = iachar(s(p:p))-digit_0
+            if( val >= 0 .and. val <= 9) then
+                v = v*10 + val ; p = p + 1
+            else
+                exit
+            end if
+        end do
+        stat = 0
+    end subroutine
+    #:endfor
+
+    elemental subroutine to_real_sp(s,v,p,stat)
+        integer, parameter :: wp    = sp
+        !> Sequentially unroll the character and get the sub integers composing the whole number, fraction and exponent
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        real(wp), intent(inout)  :: v !> Output real value
+        integer(int8), intent(out)  :: p !> last position within the string
+        integer(int8), intent(out)  :: stat !> status upon success or failure to read
+
+        ! -- Internal Variables
+        integer(kind=ikind), parameter :: nwnb = 39 !> number of whole number factors
+        integer(kind=ikind), parameter :: nfnb = 37 !> number of fractional number factors
+        integer :: e
+        real(dp), parameter :: whole_number_base(nwnb)   = [(10._dp**(nwnb-e),e=1,nwnb)]
+        real(dp), parameter :: fractional_base(nfnb)   = [(10._dp**(-e),e=1,nfnb)]
+        real(dp), parameter :: expbase(nwnb+nfnb) = [whole_number_base, fractional_base]
```

Comment:
should these be `real(dp)` or `real(sp)`?

Reply:
Using dp here is what enabled obtaining exact precision for sp, otherwise we'll comeback to the initial issue pointed out by @klausler, see this comment above for reference: https://github.com/fortran-lang/stdlib/pull/743#issuecomment-1791953430 for

Reply:
Ok. Thank you. Therefore I suggest to add a comment to prevent a future change, e.g.:
```suggestion
        ! We use dp here to obtain exact precision for sp.
        ! Otherwise errors may appear in comparison to formatted read.
        ! See  https://github.com/fortran-lang/stdlib/pull/743#issuecomment-1791953430 for more details

        real(dp), parameter :: whole_number_base(nwnb)   = [(10._dp**(nwnb-e),e=1,nwnb)]
        real(dp), parameter :: fractional_base(nfnb)   = [(10._dp**(-e),e=1,nfnb)]
        real(dp), parameter :: expbase(nwnb+nfnb) = [whole_number_base, fractional_base]
```

Reply:
Providing tests that would fail if `dp` is changed to `sp` would be useful too (if it is not the case yet).

Reply:
Added the comment in the last commit. 
The test for sp in that cause should fail as for sp the check is on a strict equality `if(abs(rel_err) > 0.0_wp)` while for dp is on a softer condition `if(abs(rel_err) > epsilon(0.0_wp))`